yt_dlp/extractor/crunchyroll.py

```diff
@@ -807,30 +828,44 @@ def _real_extract(self, url):
 
-        formats = []
+        available_formats = {}
         for stream_type, streams in get_streams('streams'):
             if stream_type not in requested_formats:
                 continue
             for stream in streams.values():
-                hardsub_lang = stream.get('hardsub_locale') or ''
-                if hardsub_lang.lower() not in requested_hardsubs:
-                    continue
-                format_id = join_nonempty(stream_type, format_field(stream, 'hardsub_locale', 'hardsub-%s'))
                 if not stream.get('url'):
                     continue
-                if stream_type.endswith('hls'):
+                hardsub_lang = stream.get('hardsub_locale') or ''
+                format_id = join_nonempty(stream_type, format_field(stream, 'hardsub_locale', 'hardsub-%s'))
+                available_formats[hardsub_lang] = (stream_type, format_id, hardsub_lang, stream['url'])
+
+        if '' in available_formats and 'all' not in requested_hardsubs:
+            full_format_langs = set(requested_hardsubs)
+            self.to_screen('To get all formats for the hardsub variants use "--extractor-args crunchyrollbeta:hardsub=all"')
+        else:
+            full_format_langs = set(map(str.lower, available_formats))
+
+        formats = []
+        for stream_type, format_id, hardsub_lang, stream_url in available_formats.values():
+            if stream_type.endswith('hls'):
+                if hardsub_lang.lower() in full_format_langs:
                     adaptive_formats = self._extract_m3u8_formats(
-                        stream['url'], display_id, 'mp4', m3u8_id=format_id,
+                        stream_url, display_id, 'mp4', m3u8_id=format_id,
                         fatal=False, note=f'Downloading {format_id} HLS manifest')
-                elif stream_type.endswith('dash'):
-                    adaptive_formats = self._extract_mpd_formats(
-                        stream['url'], display_id, mpd_id=format_id,
-                        fatal=False, note=f'Downloading {format_id} MPD manifest')
-                for f in adaptive_formats:
-                    if f.get('acodec') != 'none':
-                        f['language'] = stream_response.get('audio_locale')
+                else:
+                    adaptive_formats = (self._m3u8_meta_format(stream_url, ext='mp4', m3u8_id=format_id),)
+            elif stream_type.endswith('dash'):
+                adaptive_formats = self._extract_mpd_formats(
+                    stream_url, display_id, mpd_id=format_id,
+                    fatal=False, note=f'Downloading {format_id} MPD manifest')
+            else:
+                continue
+            for f in adaptive_formats:
+                if f.get('acodec') != 'none':
+                    f['language'] = stream_response.get('audio_locale')
+                if requested_hardsubs:
                     f['quality'] = hardsub_preference(hardsub_lang.lower())
```

Comment:
Not clear what this `if` is doing. I think it's *always* true, except possibly when the extractor arg is explicitly set to an empty string.

Also, shouldn't we set quality lower for the meta formats than others? Or does it not actually end up mattering?

Reply:
`InfoExtractor._m3u8_meta_format` already decreases the preference by 100, so we don't need to alter quality.
You are right about the `if`, that is an artefact from where `requested_hardsubs` was not a `set`. I will remove it.

Reply:
Ah, ok. Saw that a few days ago but forgot about it.

Reply:
Actually, do we really want to alter the quality? Theoretically this should be handled by the `preference` key instead, right?

Reply:
I'm not familiar enough with the ordering and preference systems in yt-dlp to know, sorry.

Reply:
I've kept it as is now, since all the other extractors do it this way as well. If it ever becomes a concern it should be tracked as another issue completely.